### PR TITLE
Convert Components to TypeScript

### PR DESCRIPTION
**Pull request by markus101**
#### Description

Converts a bunch of components to TypeScript.

### CODE REVIEW

**Review by Treycos — COMMENTED** [review]
I think I'll stop there since most comments are more or less about the same design recommendations   

In almost every components, I'd recommend extending other interfaces (native or not) if possible, to avoid duplicating types and facilitate core components updates.

A minimal component that implements these recommendations would be `TableRowCellprops` in this PR :
https://github.com/Sonarr/Sonarr/pull/7145/files

TL:DR
- Extend any native or already existing types
- Export the interface to allow other components to do the same
- Spread the whole thing in the component JSX whenever it makes sense
- Use `Pick` when you want a subset of an existing interface
- Use `FooProps['<name>']` when you want to import a prop under another name

File: frontend/src/Components/HeartRating.tsx [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
 import Tooltip from 'Components/Tooltip/Tooltip';
 import { icons, kinds, tooltipPositions } from 'Helpers/Props';
 import translate from 'Utilities/String/translate';
 import styles from './HeartRating.css';
 
-function HeartRating({ rating, votes, iconSize }) {
+interface HeartRatingProps {
+  rating: number;
+  votes?: number;
+  iconSize?: number;
```
Comment (Treycos) [inline frontend/src/Components/HeartRating.tsx:11]:
```suggestion
  iconSize?: IconProps['size'];
```

File: frontend/src/Components/Scroller/OverlayScroller.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+import { OnScroll } from './Scroller';
+import styles from './OverlayScroller.css';
+
+const SCROLLBAR_SIZE = 10;
+
+interface OverlayScrollerProps {
+  className?: string;
+  trackClassName?: string;
+  scrollTop?: number;
+  scrollDirection: unknown | unknown | unknown;
```
Comment (Treycos) [inline frontend/src/Components/Scroller/OverlayScroller.tsx:12]: Typo ?

File: frontend/src/Components/SpinnerIcon.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { FontAwesomeIconProps } from '@fortawesome/react-fontawesome';
+import React from 'react';
+import { icons } from 'Helpers/Props';
+import Icon from './Icon';
+
+interface SpinnerIconProps {
+  className?: string;
+  name: FontAwesomeIconProps['icon'];
+  spinningName?: FontAwesomeIconProps['icon'];
```
Comment (Treycos) [inline frontend/src/Components/SpinnerIcon.tsx:9]:
```suggestion
  spinningName?: IconProps['name'];
```

File: frontend/src/Components/Tooltip/Tooltip.tsx [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+import dimensions from 'Styles/Variables/dimensions';
+import { isMobile as isMobileUtil } from 'Utilities/browser';
+import styles from './Tooltip.css';
+
+export interface TooltipProps {
+  className?: string;
+  bodyClassName?: string;
+  anchor: React.ReactNode;
+  tooltip: string | React.ReactNode;
+  kind?: Extract<(typeof kinds.all)[number], keyof typeof styles>;
```
Comment (Treycos) [inline frontend/src/Components/Tooltip/Tooltip.tsx:21]: I didn't anticipate the type `(typeof kinds.all)[number]` would be used so much (same for sizes)

Exporting it from `Kinds` under a simpler name might be a good idea now

File: frontend/src/Components/Tooltip/Tooltip.tsx [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    } else if (windowWidth >= parseInt(dimensions.breakpointSmall)) {
+      return 500;
+    }
+
+    return 450;
+  }, []);
+
+  const computeMaxSize = useCallback(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    (data: any) => {
```
Comment (Treycos) [inline frontend/src/Components/Tooltip/Tooltip.tsx:89]: Only type I could get working is this this `react-popper` doesn't export much from `popper`

Not great 🙁 

```suggestion
  const computeMaxSize = useCallback<
    NonNullable<
      NonNullable<
        NonNullable<PopperProps['modifiers']>['computeMaxHeight']
      >['fn']
    >
  >(data) => {
```

File: frontend/src/Components/Tooltip/Tooltip.tsx (hunk 4 above)
Comment (markus101) [reply frontend/src/Components/Tooltip/Tooltip.tsx:21]: Good call, did the same for the others. Ended up extending the types a bit to avoid issues with them being treated as strings in some places.

File: frontend/src/Components/Tooltip/Tooltip.tsx (hunk 5 above)
Comment (markus101) [reply frontend/src/Components/Tooltip/Tooltip.tsx:89]: Yeah, saw there wasn't much exported and gave up. I ran into issues with this change and reverted back, but not too worried, eventually popper will get updated/replaced and it'll improve.

**Review by markus101 — COMMENTED** [review]
Thanks for the feedback, addressed most of the comments/added some replies.

File: frontend/src/Components/Portal.tsx [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+interface PortalProps {
+  children: React.ReactNode;
+  target?: Element | DocumentFragment;
+}
```
Comment (Treycos) [inline frontend/src/Components/Portal.tsx:7]:
```suggestion
interface PortalProps {
  children: Parameters<typeof ReactDOM.createPortal>[0];
  target?: Parameters<typeof ReactDOM.createPortal>[1];
}
```

File: frontend/src/Components/Portal.tsx [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+function Portal(props: PortalProps) {
+  const { children, target = document.getElementById('portal-root') } = props;
```
Comment (Treycos) [inline frontend/src/Components/Portal.tsx:10]: You might want to move the `getElementById` call above the component to avoid calling it for every portal

File: frontend/src/Components/SpinnerIcon.tsx [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
```
Comment (Treycos) [inline frontend/src/Components/SpinnerIcon.tsx:1]: Did the same refactor on another branch
Here's my version

```suggestion
import React from 'react';
import { icons } from 'Helpers/Props';
import Icon, { IconProps } from './Icon';

export interface SpinnerIconProps extends IconProps {
  spinningName?: IconProps['name'];
  isSpinning: Required<IconProps['isSpinning']>;
}

export default function SpinnerIcon({
  name,
  spinningName = icons.SPINNER,
  ...otherProps
}: SpinnerIconProps) {
  return (
    <Icon
      name={(otherProps.isSpinning && spinningName) || name}
      {...otherProps}
    />
  );
}
```

File: frontend/src/Helpers/Props/align.ts [hunk 9]
```diff
@@ -3,3 +3,5 @@ export const CENTER = 'center';
 export const RIGHT = 'right';
 
 export const all = [LEFT, CENTER, RIGHT];
+
+export type Align = ('left' | 'center' | 'right') & (typeof all)[number];
```
Comment (Treycos) [inline frontend/src/Helpers/Props/align.ts:7]: Is the type redundancy intentional ?

These would be the same type

```suggestion
export type Align = 'left' | 'center' | 'right;
```

```suggestion
export type Align = (typeof all)[number];
```

Comment (Treycos) [reply frontend/src/Helpers/Props/align.ts:7]: Same question for Kinds, Sizes, directions... etc

File: frontend/src/Components/Tooltip/Popover.tsx [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import Tooltip, { TooltipProps } from './Tooltip';
+import styles from './Popover.css';
+
+interface PopoverProps extends Omit<TooltipProps, 'tooltip' | 'bodyClassName'> {
+  title: string;
+  body: string | React.ReactNode;
```
Comment (Treycos) [inline frontend/src/Components/Tooltip/Popover.tsx:7]: Redundant typing

```suggestion
  body: React.ReactNode;
```

File: frontend/src/Components/Scroller/OverlayScroller.tsx [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  autoHide: boolean;
+  autoScroll: boolean;
+  children?: React.ReactNode;
+  onScroll?: (payload: OnScroll) => void;
+}
+
+interface ScrollbarTrackProps {
+  style: React.CSSProperties;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  props: any;
```
[the remark this answers is not in the document]
Comment (markus101) [reply frontend/src/Components/Scroller/OverlayScroller.tsx:23]: Not sure if I replied to the wrong thing, but yeah that definitely works 🤔

File: frontend/src/Components/SpinnerIcon.tsx (hunk 8 above)
Comment (markus101) [reply frontend/src/Components/SpinnerIcon.tsx:1]: Totally spaced that you did this one already, making a small tweak to not use `otherProps` inside of the component at the cost of a couple extra lines 😄

File: frontend/src/Helpers/Props/align.ts (hunk 9 above)
Comment (markus101) [reply frontend/src/Helpers/Props/align.ts:7]: Yes, the indexed strings were causing issues in some places so I extended the type, but forgot how TS resolves the types/strings in that scenario. I'll clean them up

File: src/NzbDrone.Core/Localization/Core/es.json [hunk 12]
```diff
@@ -1,2108 +1,2108 @@
 {
```
Comment (Treycos) [inline src/NzbDrone.Core/Localization/Core/es.json:1]: Are all the formatting changes wanted or is it your formater derping 🤔

Comment (markus101) [reply src/NzbDrone.Core/Localization/Core/es.json:1]: Formatter went rogue 😞
